plugins/onepad/SDL/joystick.h

```diff
@@ -32,59 +32,31 @@
 class JoystickInfo : GamePad
 {
 public:
-    JoystickInfo()
-        : GamePad()
-        , joy(nullptr)
-    {
-        haptic = nullptr;
-        first = true;
-        memset(effects, 0, sizeof(effects));
-        memset(effects_id, 0, sizeof(effects_id));
-    }
+    JoystickInfo(int id);
+    ~JoystickInfo();
 
-    ~JoystickInfo()
-    {
-        Destroy();
-    }
+    JoystickInfo(const JoystickInfo &) = delete;            // copy constructor
+    JoystickInfo &operator=(const JoystickInfo &) = delete; // assignment
 
-    JoystickInfo(const JoystickInfo &);            // copy constructor
-    JoystickInfo &operator=(const JoystickInfo &); // assignment
 
-    void Destroy();
     // opens handles to all possible joysticks
-    static void EnumerateJoysticks(vector<GamePad *> &vjoysticks);
+    static void EnumerateJoysticks(std::vector<std::unique_ptr<GamePad>> &vjoysticks);
 
-    void Rumble(int type, int pad);
-
-    bool Init(int id); // opens a handle and gets information
+    void Rumble(unsigned type, unsigned pad);
```

Comment:
I think override/final is enough, virtual is a bit redundant if either is specified.